### ISSUE TITLE
0010780:
MessageBox "Session invalid" occurs if Browser is closed without logout in the last Session

### ISSUE THREAD

**Issue by Gloirin** [tine20/Tine-2.0-Open-Source-Groupware-and-CRM#5336]
**Reported by ravermeister on 27 Jan 2015 11:17**

**Version:** Koriander (2014.09.7)

Hi, when an Old was not closed Correctly a Messagebox appears,
saying that the Session is invalid.
[img]http://www.rimkus.de.be/devel/tine20/login/login_error.png[img]
See the German Thread about it:
https://www.tine20.org/forum/viewtopic.php?f=12&amp;t=16992&amp;start=10

**Steps to reproduce:** 1. Open tine20
2. Login
3. Close Browser
4. Reopen Browser
5. Open tine20

**Additional information:** A Trace Log is attached, or available here:
[url]http://pastebin.com/Rix8a7Mm[/url]

**Comment by Gloirin** [comment]
Related to #1939

**Comment by Gloirin** [comment]
Related to #5464

**Comment by Gloirin** [comment]
**Comment posted by ravermeister on 27 Jan 2015 11:20**

screenshot of messagebox: http://www.rimkus.it/devel/tine20/login/login_error.png

Trace Log: http://pastebin.com/Rix8a7Mm

**Comment by Gloirin** [comment]
**Comment posted by ravermeister on 30 Jan 2015 08:11**

Forum Thread: https://www.tine20.org/forum/viewtopic.php?f=12&amp;t=16992&amp;start=10

**Comment by Gloirin** [comment]
**Comment posted by ravermeister on 8 Feb 2015 14:35**

Confirmed Bug, by mspahn:
https://www.tine20.org/forum/viewtopic.php?f=12&amp;t=16992&amp;start=10#p60468

**Comment by Gloirin** [comment]
**Comment posted by ravermeister on 1 Mar 2015 13:07**

Seems fixed in Version: Koriander (2014.09.8)

**Comment by Gloirin** [comment]
**Comment posted by ravermeister on 27 Mar 2015 09:45**

in the Current version 2014.09.9 this bug appears again!

**Comment by Gloirin** [comment]
**Comment posted by pschuele on 29 Mar 2015 20:24**

hi, thanks for reporting. I&#039;ll have a look.

this might be related to #3726

**Comment by Gloirin** [comment]
**Comment posted by pschuele on 29 Mar 2015 20:41**

http://gerrit.tine20.com/customers/#/c/1775/

**Comment by Gloirin** [comment]
**Comment posted by ravermeister on 29 Mar 2015 22:21**

could bring some clarification in this? 
why is it related to the mailto feature request?
is it a bug introduced with this feature?
is it now fixed?

**Comment by Gloirin** [comment]
**Comment posted by pschuele on 30 Mar 2015 09:41**

#3726 moved the tine20 registry to the browsers local storage. this changed lots of things and introduced this bug because the registry persists when the browser is closed. we need to implement a new check in the client if the current session is still valid to show the login box (if not valid).

this will be fixed in the next service release.

**Comment by Gloirin** [comment]
**Comment posted by pschuele on 29 Apr 2015 08:57**

some attempts fixing this: http://gerrit.tine20.com/customers/#/c/1775/3 (should be no longer needed)

**Comment by Gloirin** [comment]
**Comment posted by ravermeister on 30 Apr 2015 18:23**

in the current version 2014.09.10
the problem is even worse,

after step 4, the whole tine crashes means
I get a half loaded screen and after a while firefox shows a message
that a js file is slowing down the page, and ask me to terminate

**Comment by Gloirin** [comment]
**Comment posted by mspahn on 1 May 2015 12:47**

Chrome and Firefox does not equally handle that.

The implementation needs to be reviewed for Firefox.

**Comment by Gloirin** [comment]
**Comment posted by ravermeister on 1 May 2015 13:29**

I&#039;ve just tested with chrome the behaviour is similar, but slightly different:

when you open tine after step 4, 
tine loads as if I were logged in, 
then suddenly stops and loads the login page.

I think Firefox does the Same, but crashes before loading again the login page,
because Palemoon (firefox clone) http://www.palemoon.org/ 

acts like chrome, but takes about 30 seconds from the broken loaded view, to the login page.

hope that helps...

**Comment by Gloirin** [comment]
**Comment posted by pschuele on 4 May 2015 07:59**

hm, i could not reproduce the problems.

is the session already gone on the server? i still had a session on the server and it could be continued without failure (Chromium + Firefox).

**Comment by Gloirin** [comment]
**Comment posted by weberho on 4 May 2015 08:03**

Can not reproduce this problem on several installations. Are there any interesting new logs available?

**Comment by Gloirin** [comment]
**Comment posted by ravermeister on 4 May 2015 14:20**

Hi, So after playing about 3 hours with Config I can say the following:

- it has nothing to do with the redis/memcache settings 
  (I tested with redis and File)
- it is always reproducable
- tine seems to loop
  --&gt; when you wait long enough, you get an error message saying
       max_user_connections limit is reached.


I have attached my Config, and the Debug log for step 5
see 
2015.05.04.config.inc.php
2015.05.04.tine.error.log

for me the log file indicates a loop. 
you can see the unused DB connections are growing during loading the tine20 page.
e.g.:  mysqladmin processlist:
+-------+-------+-----------+------------+---------+------+-------+------------------+
| Id    | User  | Host      | db         | Command | Time | State | Info             |
+-------+-------+-----------+------------+---------+------+-------+------------------+
| 31510 | raver | localhost | raver_tine | Sleep   | 16   |       |                  |
| 31947 | raver | localhost | raver_tine | Sleep   | 27   |       |                  |
| 32333 | raver | localhost | raver_tine | Sleep   | 2    |       |                  |
| 32345 | raver | localhost | raver_tine | Sleep   | 2    |       |                  |
| 32349 | raver | localhost | raver_tine | Sleep   | 2    |       |                  |
| 32351 | raver | localhost | raver_tine | Sleep   | 2    |       |                  |
| 32352 | raver | localhost | raver_tine | Sleep   | 2    |       |                  |
| 32353 | raver | localhost | raver_tine | Sleep   | 2    |       |                  |
| 32361 | raver | localhost | raver_tine | Sleep   | 1    |       |                  |
| 32362 | raver | localhost | raver_tine | Sleep   | 1    |       |                  |
| 32364 | raver | localhost | raver_tine | Sleep   | 1    |       |                  |
| 32365 | raver | localhost | raver_tine | Sleep   | 1    |       |                  |
| 32368 | raver | localhost | raver_tine | Sleep   | 1    |       |                  |
| 32371 | raver | localhost | raver_tine | Sleep   | 1    |       |                  |
| 32372 | raver | localhost | raver_tine | Sleep   | 1    |       |                  |
| 32378 | raver | localhost | raver_tine | Sleep   | 1    |       |                  |
| 32383 | raver | localhost | raver_tine | Sleep   | 0    |       |                  |
| 32384 | raver | localhost | raver_tine | Sleep   | 0    |       |                  |
| 32387 | raver | localhost |            | Query   | 0    |       | show processlist |
+-------+-------+-----------+------------+---------+------+-------+------------------+

until this Error is thrown:
Zend_Db_Adapter_Exception -&gt; SQLSTATE[HY000] [1226] User &#039;raver&#039; has exceeded the &#039;max_user_connections&#039; resource (current value: 20)

So I think here is definately an Error


P.S.: The firefox message shows this:

Ein Skript auf dieser Seite ist eventuell beschäftigt oder es antwortet nicht mehr. 
Sie können das Skript jetzt stoppen, 
im Debugger öffnen oder weiter ausführen.
Skript: http://mail.rimkus.it/index.php?method=Tinebase.getJsFiles:48

My Hardware Specs are:
PHP:
----
PHP 5.6.3 (cli) (built: Nov 14 2014 20:21:33)
Copyright (c) 1997-2014 The PHP Group
Zend Engine v2.6.0, Copyright (c) 1998-2014 Zend Technologies
    with Zend OPcache v7.0.4-dev, Copyright (c) 1999-2014, by Zend Technologies


Mysql:
-----
Welcome to the MySQL monitor.  Commands end with ; or \g.
Your MySQL connection id is 33725
Server version: 5.1.73-log Source distribution

Copyright (c) 2000, 2013, Oracle and/or its affiliates. All rights reserved.

Oracle is a registered trademark of Oracle Corporation and/or its
affiliates. Other names may be trademarks of their respective
owners.


Apache:
-------
Server version: Apache/2.2.15 (Unix)
Server built:   Oct 16 2014 14:48:21


Hope that helps, let me know if I can do more

**Comment by Gloirin** [comment]
**Comment posted by weberho on 4 May 2015 14:46**

&quot;User &#039;raver&#039; has exceeded the &#039;max_user_connections&#039; resource (current value: 20)&quot; is very interesting.

* Is this installation hosted on a shared-host environment?

* Are other applications creating connections with the same user &#039;raver&#039;?

* How many users are accessing tine20 at the same time (including ActiveSync devices)

IMHO 20 connections are not enough; I have experienced several situations where connections were kept open for a longer time and even 100+ connections were not enough (I could never find the reason for that).

**Comment by Gloirin** [comment]
**Comment posted by ravermeister on 4 May 2015 15:22**

&quot;User &#039;raver&#039; has exceeded the &#039;max_user_connections&#039; resource (current value: 20)&quot; is very interesting.
* Is this installation hosted on a shared-host environment? --&gt; yes https://uberspace.de/

* Are other applications creating connections with the same user &#039;raver&#039;? no

* How many users are accessing tine20 at the same time (including ActiveSync devices) average are 3 to 4

IMHO 20 connections are not enough; I have experienced several situations where connections were kept open for a longer time and even 100+ connections were not enough (I could never find the reason for that).  
--&gt; you are kidding me! 100+ are not enough? and you still think threre is no bug in the source code?
come on you can not be serious.
DB connections are reused when they are in &#039;sleep&#039;

See the answer from my Hosting Support: http://pastebin.com/Wd0rkBK4


and nevertheless the Problem is NOT present in 2015.09.08!!!
so something MUST have changed

**Comment by Gloirin** [comment]
**Comment posted by weberho on 4 May 2015 16:02**

I agree, there must be something wrong. Administering several Linux-servers I see sometimes lots off sleeping connections to the DB-server coming from Apache/PHP; they are kept open even though other users might be in need of connections.

Even when the problem is caused by tine, you can only expect it to be running when tine can connect to the server cleanly.

My impression is that the tine20 mail-sync consumes much resources and keep open connections for a very long time.

**Comment by Gloirin** [comment]
**Comment posted by ravermeister on 4 May 2015 18:19**

...I think this: https://forge.tine20.org/view.php?id=10780#c15624
 could still make some problems...

but to be honest, I simply have no Idea. 
if I can test something let me know...

**Comment by Gloirin** [comment]
**Comment posted by pschuele on 14 May 2015 09:52**

i finally could reproduce the error by deleting the sessions on the server. i could debug the js code in chrome afterwards and detected an inifinite loop in the clearRegistry function. this should be fixed now, I&#039;ll push the fix to git later this day.

**Comment by Gloirin** [comment]
**Comment posted by pschuele on 15 May 2015 08:46**

regarding the max_connections: this might be sql connections that are kept open (sleeping) by the ActiveSync ping request which is a long running request. maybe it would be better that those connections are closed and reopened if needed ... but this should be moved to a separate issue.

**Comment by Gloirin** [comment]
**Comment posted by ravermeister on 15 May 2015 16:48**

Great! Thanks :)
will try it out as soon it is released.

regarding the other Issue, 
is it okay to open a new Bug and link to this one?
or should I copy all Information from here?

**Comment by Gloirin** [comment]
**Comment posted by pschuele on 18 May 2015 08:36**

&gt; is it okay to open a new Bug and link to this one?
&gt; or should I copy all Information from here?

just open a new one, I&#039;ll link it to this one.

edit: #11040 is the new bugreport.

**Comment by Gloirin** [comment]
**Comment posted by ravermeister on 29 May 2015 15:36**

The problem is still present in 2014.09.12!
the only difference is that tine will abort much faster,
but tine still tries to restore the invalid session show a broken E-mail View
for about a Second or so, and the loads the Login screen...

**Comment by Gloirin** [comment]
**Comment posted by ravermeister on 29 May 2015 16:03**

I&#039;ve added a new Trace Log, 
altrenatively it is available here: http://pastebin.com/jfTwcTvS


I&#039;ve cleaned the log file before I reproduced the Error, 
I think this is a complete log, because I find to lines of:

METHOD: Tinebase_Server_Json::Tinebase.getAllRegistryData / TIME: 40ms /  Memory usage: 4.5 MB /  Real patch cache size: 65474 / PID: 18421

and I guess the first is the invalid session.

**Comment by Gloirin** [comment]
**Comment posted by ravermeister on 29 May 2015 16:06**

please let me know if you need any additional information

**Comment by Gloirin** [comment]
**Comment posted by ravermeister on 10 Aug 2015 12:26**

The problem is still present with the curernt release: Stable: Elena (2015.07.2)
any news on this?

**Comment by Gloirin** [comment]
**Comment posted by pschuele on 11 Aug 2015 14:15**

hi raver,

we did a little brainstorming on this and how it should be solved. we would need an additional request, that checks if a valid session exists as the client does not know this. this needs to be done in each window when it is rendered.

but we currently have no exact idea when this is going to be implemented.

**Comment by Gloirin** [comment]
**Comment posted by ravermeister on 13 Aug 2015 22:34**

I tested with the Application Modal setting 
means all Dialogs open in the Same Browser Window
and  I could not reproduce it with the current Stable: Elena (2015.07.3)
but with the versions before

maybe this fix has somehow solved it?
https://forge.tine20.org/view.php?id=11198

**Comment by Gloirin** [comment]
**Comment posted by pschuele on 14 Aug 2015 08:07**

ah, i think another fix (http://git.tine20.org/git?p=tine20;a=commit;h=46d36434496d5a72c3bc95b68c5644ddff8edd25) should have solved that.

thanks for testing :)

I&#039;ll close the issue.

**Comment by Gloirin** [comment]
**Comment posted by pschuele on 21 Sep 2015 11:52**

please describe the current problem here. is it still the same as in the description? i don&#039;t think that we have a major problem here. only the session detection could be improved to remove the need of a message box.

**Comment by Gloirin** [comment]
**Comment posted by ravermeister on 21 Sep 2015 14:22**

Hi,
yes it is still the same as in the description:
Settings:
- redis enabled
- custom tmp folder
- application modal windows

1. Login into tine
2. Close Browser without logout
3. navigate to tine login page

- a Message Dialog appears for a second or so saying Session is invalid, and page is reloaded
- now the login panel appears.	

a new error log wich shows the issue is available here:
https://www.rimkus.it/stuff/tine_error/tine.log.err

**Comment by Gloirin** [comment]
**Comment posted by ravermeister on 22 Sep 2015 06:10**

I think this are the interesting log entries:
e0710 b9e87 -- none -- - 2015-09-21T13:39:30+00:00 INFO (6): Tinebase_Server_Json::_handle::251 is JSON request. method: Felamimail.searchFolders
e0710 b9e87 -- none -- - 2015-09-21T13:39:30+00:00 INFO (6): Tinebase_Server_Json::_checkJsonKey::402 Attempt to request a privileged Json-API method (Felamimail.searchFolders) without authorisation from &quot;130.180.102.78&quot;. (session timeout?)
e0710 b9e87 -- none -- - 2015-09-21T13:39:30+00:00 WARN (4): Tinebase_Server_Json::_handleException::325 Tinebase_Exception_AccessDenied -&gt; Not Authorised
e0710 b9e87 -- none -- - 2015-09-21T13:39:30+00:00 ERR (3): Tinebase_Exception::log::104 Tinebase_Exception_AccessDenied -&gt; Not Authorised
e0710 b9e87 -- none -- - 2015-09-21T13:39:30+00:00 INFO (6): index.php (25) METHOD: Tinebase_Server_Json:: / TIME: 25ms /  Memory usage: 2.5 MB /  Real patch cache size: 65469 / PID: 3839


if you need more information let me know :)

**Comment by Gloirin** [comment]
**Comment posted by ravermeister on 2 Oct 2015 08:24**

any news according this issue?
everytime when I provide the requested information, 
it gets very calm here....

I can Screen-record the problem and put it on Youtube if you like :)

**Comment by Gloirin** [comment]
**Comment posted by pschuele on 2 Oct 2015 09:08**

i think the only way to prevent this behavior (that we don&#039;t classify as a major issue) would be to introduce a valid session detection when the mainscreen is rendered.

I&#039;ll let you know when there are any news on this, but it may take a while.

**Comment by Gloirin** [comment]
**Comment posted by ravermeister on 5 Oct 2015 12:47**

thanks but I still don&#039;t get it, 
because I set my Session Cookie lifetime to 0 in tine, which means there is NO Session Cookie after Closing the Browser. 
(I have verified that!)

and therefore $_SESSION vars are NOT available, so I do not see the problem of Session detection. 

But fine, I&#039;ll wait....
;)

**Comment by Gloirin** [comment]
**Comment posted by cweiss on 15 Feb 2017 07:41**

I don&#039;t understand this report. Is there a login loop or some misbehaviour?

Otherwise it&#039;s the desired behaviour. If you open tine20 and your session is invalid you get this message and you are redirected to login after confirming the message.

It would be a feature request to silently discard the problem and to show login box directly.

**Comment by Gloirin** [comment]
**Comment posted by ravermeister on 15 Feb 2017 15:12**

I think it is a Mis behaviour. Because in earlier Versions this did not happened.

1. Login into Tine.
2. Completely Close your Browser WITHOUT logging out in tine20
3. Wait as long as you want (even a day)
4. enter tine login page.

now Tine thinks I&#039;m already signed in and tries to load E-Mail Tab/Calendar Tab etc.
in a few mili seconds following messageBox appears: 
http://www.rimkus.it/devel/tine20/login/login_error.png

then tine reloads the page and I see the login panel. 
in earlier Versions, no message was shown at all, and I got directly to the login box exactly as if you would press the &quot;logout&quot; link.

If you need further information, please let me know

kind regards 
Ravermeister

**Comment by Gloirin** [comment]
**Comment posted by cweiss on 16 Feb 2017 11:56**

Thanks, its clear now.

It&#039;s the desired behaviour. It was just a problem, that the user wasn&#039;t able to press the ok button as the reload started immediately. 

The general idea is, that some day we detect the invalid session and show a login-page on every open window. Once the user reauthenticated the login vanishes from all windows. But I&#039;m not sure if we&#039;ll ever implement this. There are also good reasons to close all other windows an redirect to login.

You might want to open a feature request to skip the confirm dialog.

But we found an other problem in this area. If you have an SSO adapter, the box appears after the SSO login and the user has to login twice. We fixed this bug and will close this ticket with it.
